### PR TITLE
Generic model tags

### PR DESCRIPTION
**Pull request by DashDeipayan**
Added generic models to create, add and remove tags

### CODE REVIEW

File: middlewares/validators/tags.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+const Joi = require("joi");
+const { TYPES } = require("../../constants/tags");
+
+const createTag = async (req, res, next) => {
```
Comment (alishvagadiya) [inline middlewares/validators/tags.js:4]: The function body and name is not matching,

In the function body, there is post-data validation, and names give hints like adding or creating tags,

better name can be  validAddTag or validTagBody,

**Review by alishvagadiya — CHANGES_REQUESTED** [review]

File: middlewares/validators/tags.js (hunk 1 above)
Comment (DashDeipayan) [reply middlewares/validators/tags.js]: fixed it

File: controllers/tags.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+const deleteTag = async (req, res) => {
+  try {
+    const { id } = await TagModel.deleteTag(req.params.tagid);
+    return res.json({
+      message: "Tag Deleted successfully!",
+      id,
+    });
+  } catch (err) {
+    logger.error(`Error while creating new tag: ${err}`);
```
Comment (heyrandhir) [inline controllers/tags.js:40]: the error message can be changed. 
```suggestion
	    logger.error(`Error while deleting the tag: ${err}`);
```

File: controllers/tags.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * @param res {Object} - Express response object
+ */
+const getAllTags = async (req, res) => {
+  try {
+    const { allTags } = await TagModel.getAllTags();
+    return res.json({
+      allTags,
+    });
+  } catch (err) {
+    logger.error(`Error while creating new tag: ${err}`);
```
Comment (heyrandhir) [inline controllers/tags.js:58]: the error message can be changed. 
```suggestion
	    logger.error(`Error while fetching all the tag: ${err}`);
```

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: controllers/tags.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * @param res {Object} - Express response object
+ */
+
+const addTag = async (req, res) => {
+  try {
+    const { id, tagData } = await TagModel.addTag(req.body);
+
+    return res.json({
+      message: "Tag created successfully!",
+      stock: tagData,
```
Comment (vinayak-trivedi) [inline controllers/tags.js:16]: any reasons to name this as stock?

File: routes/tags.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+const express = require("express");
+
+const { addTag, deleteTag, getAllTags } = require("../controllers/tags");
+const { validTagBody } = require("../middlewares/validators/tags");
+
+const router = express.Router();
+
+router.post("/", validTagBody, addTag);
+router.delete("/:tagid", deleteTag);
```
Comment (vinayak-trivedi) [inline routes/tags.js:9]: we don't want everyone to add or delete tags, can you authenticate for the superuser role here

File: controllers/tags.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Creates a new tag
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addTag = async (req, res) => {
+  try {
+    const { id, tagData } = await TagModel.addTag(req.body);
```
Comment (vinayak-trivedi) [inline controllers/tags.js:12]: Are we sending createdBy in the body only? Can we use the authenticate middleware here to get the user data and put that in the createdBy field?

**Review by vinayak-trivedi — CHANGES_REQUESTED** [review]
Thanks for the PR @DashDeipayan 
Suggested some changes please have a look

File: controllers/tags.js (hunk 4 above)
Comment (DashDeipayan) [reply controllers/tags.js:16]: Oh, missed it. will change it to data

File: routes/tags.js (hunk 5 above)
Comment (DashDeipayan) [reply routes/tags.js:9]: sure, will add it

File: controllers/tags.js (hunk 6 above)
Comment (DashDeipayan) [reply controllers/tags.js:13]: done

**Review by vinayak-trivedi — APPROVED** [review]
LGTM, just one NIT comment, you can have a look at it

File: controllers/tags.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ * Creates a new tag
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addTag = async (req, res) => {
+  try {
+    req.body.createdby = req.userData.id;
+    const { id, tagData } = await TagModel.addTag(req.body);
```
[the remark this answers is not in the document]
Comment (DashDeipayan) [reply controllers/tags.js:13]: Thanks Vinayak, that's better definitely

**Review by alishvagadiya — DISMISSED** [review]

**Review by heyrandhir — DISMISSED** [review]
LGTM

File: controllers/tags.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Creates a new tag
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+
+const addTag = async (req, res) => {
+  try {
+    const { id, tagData } = await TagModel.addTag({ ...req.body, createdby: req.userData.id, date: new Date() });
```
Comment (ivinayakg) [inline controllers/tags.js:12]: date is not in the epoch

File: models/tags.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ *
+ * @param tagData { Object }: Data of the tag
+ * @returns {Promise<tagModel|Object>}
+ */
+
+const addTag = async (tagData) => {
+  try {
+    tagData.type = tagData.type.toUpperCase();
```
Comment (ivinayakg) [inline models/tags.js:12]: check wether there is already a tag with the same name and type present in the database

**Review by ivinayakg — CHANGES_REQUESTED** [review]

**Review by alishvagadiya — CHANGES_REQUESTED** [review]

File: controllers/tags.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    return res.json({
+      tags,
+    });
+  } catch (err) {
+    logger.error(`Error while fetching the tags: ${err}`);
+    return res.boom.badImplementation("An internal server error occurred");
+  }
+};
+
+module.exports = {
```
Comment (alishvagadiya) [inline controllers/tags.js:86]: Return response formate is changing at all control functions. Make it consistent.

**Review by alishvagadiya — CHANGES_REQUESTED** [review]

File: models/tags.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    throw err;
+  }
+};
+
+/**
+ *
+ * @returns {Promise<tagModel|Array>}: returns an array of all the tags with its meta data
+ */
+
+const getAllTags = async () => {
```
Comment (alishvagadiya) [inline models/tags.js:49]: add try-catch block

File: models/tags.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * @param tagData { Object }: Data of the tag
+ * @returns {Promise<tagModel|Object>}
+ */
+
+const addTag = async (tagData) => {
+  try {
+    tagData.type = tagData.type.toUpperCase();
+    tagData.name = tagData.name.toUpperCase();
+
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply models/tags.js:14]: makes sense let me do that

File: models/tags.js [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+  }
+};
+
+/**
+ *
+ * @returns {Promise<tagModel|Array>}: returns an array of all the tags with its meta data
+ */
+
+const getAllTags = async () => {
+  const data = await tagModel.get();
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply models/tags.js]: I don't know why depaiyan did it, but this is the only data that is there in the db, let me use the spread syntax, to get this data

File: models/tags.js [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      reason: doc.data().reason,
+    };
+    allTags.push(tag);
+  });
+  return { allTags };
+};
+
+const getTagByType = async (type) => {
+  try {
+    const data = await tagModel.where("type", "==", type).get();
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply models/tags.js:75]: resolving it, doing it with the same approach mentioned above

File: models/tags.js [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+const addTag = async (tagData) => {
+  try {
+    tagData.type = tagData.type.toUpperCase();
+    tagData.name = tagData.name.toUpperCase();
+
+    const alreadyIsTag = await tagModel.where("name", "==", tagData.name).limit(1).get();
+    if (!alreadyIsTag.empty) {
+      return { id: "", tagData };
```
[the remark this answers is not in the document]
Comment (vinayak-trivedi) [reply models/tags.js]: I guess, depaiyan has taken care of this in his previous commit, please have a look and let me know if any further changes are required

File: controllers/tags.js (hunk 10 above)
Comment (vinayak-trivedi) [reply controllers/tags.js:88]: I think it is being taken care of in this commit , let me know if you still need some changes https://github.com/Real-Dev-Squad/website-backend/pull/813/commits/30328a69b5051309cf86063b3ac48bbd36140e6b

**Review by alishvagadiya — APPROVED** [review]

File: controllers/tags.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    if (id) {
+      return res.json({
+        message,
+        data: { tag: tagData, id },
+      });
+    } else {
+      return res.status(400).json({
+        message,
+        data: { tag: tagData },
+      });
```
Comment (ankushdharkar) [inline controllers/tags.js:27]: Nit: If-Else can be skipped here. Just use early returns.

File: models/tags.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+      allTags.push(tag);
+    });
+    return { allTags };
+  } catch (err) {
+    logger.error("error getting tags", err);
+    throw err;
+  }
+};
+
+const getTagByType = async (type) => {
```
Comment (ankushdharkar) [inline models/tags.js:69]: Should this be `getTagsByType`

**Pull request merged**